### PR TITLE
Add support for Azure multicontainer

### PR DESCRIPTION
**Pull request by dkarlovi**
#446

---
This change is [Reviewable](https://reviewable.io/reviews/knplabs/gaufrette/486)

### CODE REVIEW

**Comment by dkarlovi** [comment]
@NiR- do you think this direction is OK?

**Comment by dkarlovi** [comment]
@NiR- bump.

**Comment by NiR-** [comment]
Sorry, I was in vacation last week. I'll check it today ;)

**Comment by dkarlovi** [comment]
Sorry to bump, it's not urgent or anything, just don't want it to get lost. Hope you had a great vacation. :+1:

File: tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?php
```
Comment (NiR-) [inline tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist]: We're deprecating this way of running functional tests. Could you rather implement `setUp` / `tearDow` methods to initialize the filesystem in your Test please? [Here's an example](https://github.com/KnpLabs/Gaufrette/blob/master/tests/Gaufrette/Functional/Adapter/AwsS3Test.php#L23).

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 2]
```diff
@@ -403,4 +481,26 @@ private function guessContentType($content)
...
+    /**
+     * @param string $key
+     *
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    private function tokenizeKey($key)
+    {
+        $containerName = $this->containerName;
+        if (true === $this->multiContainerMode) {
```
Comment (NiR-) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: I would rather invert the conditional ` true === $this->multiContainerMode` to early return a value when it's disabled. It would make it a bit more readable.

File: tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace Gaufrette\Functional\Adapter;
+
+/**
+ * Class AzureMultiContainerBlobStorageTest
+ * @group AzureBlobStorage
+ * @group AzureMultiContainerBlobStorage
+ */
+class AzureMultiContainerBlobStorageTest extends FunctionalTestCase
```
Comment (NiR-) [inline tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php]: As you're overriding test cases, functional test for `multiContainerMode` disabled are not run. I think it would be better to instantiate two filesystems here and create new test cases for multi container mode.

**Comment by dkarlovi** [comment]
@NiR- thanks! :+1:

File: tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php (hunk 3 above)
Comment (dkarlovi) [inline tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php]: I don't understand this one. I've added a new test suite, haven't disabled the existing one. By default, both will be run (and I have run both test sets).

Comment (dkarlovi) [inline tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php]: $ vendor/bin/phpunit --group AzureBlobStorage --testdox
    PHPUnit 3.7.38 by Sebastian Bergmann.
    
    Configuration read from /home/dado/Development/Gaufrette/phpunit.xml.dist
    
    Gaufrette\Functional\Adapter\AzureBlobStorage
    [x] Should write and read
    [x] Should update file content
    [x] Should check if file exists
    [x] Should get mtime
    [x] Should fail when try mtime for key which does not exist
    [x] Should rename file
    [x] Should delete file
    [x] Should fetch keys
    [x] Should work with hidden files
    [x] Should keep file object in register
    [x] Should wrtie to same file
    
    Gaufrette\Functional\Adapter\AzureMultiContainerBlobStorage
    [x] Should write and read
    [x] Should update file content
    [x] Should check if file exists
    [x] Should get mtime
    [x] Should fail when try mtime for key which does not exist
    [x] Should rename file
    [x] Should delete file
    [x] Should fetch keys
    [x] Should work with hidden files
    [x] Should keep file object in register
    [x] Should wrtie to same file

File: tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist (hunk 1 above)
Comment (dkarlovi) [inline tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist]: Done.

File: tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php (hunk 3 above)
Comment (dkarlovi) [inline tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php]: Everything still passes with the last commit.

Comment (NiR-) [inline tests/Gaufrette/Functional/Adapter/AzureMultiContainerBlobStorageTest.php]: Haa indeed, sorry :+1:

File: tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist (hunk 1 above)
Comment (NiR-) [inline tests/Gaufrette/Functional/adapters/AzureMultiContainerBlobStorage.php.dist]: Thank you :) I will set up Azure Storage to ensure tests against code landing on our repository tests are are run.

**Comment by NiR-** [comment]
@dkarlovi :+1: Could you squash "wip" commits, and in the meantime, add a note in the `doc/` about this new feature please? Would be glad to see it merged :)

**Comment by dkarlovi** [comment]
Will do, thanks.

BTW I've noticed one problem still, will need to fix it too. Azure SDK guys keep changing the interface, for example they've renamed `setBlobContentType()` to `setContentType()` (which is used if `detectContentType=true`).

Adapters should really get their own composer.json so they can lock the SDK dependency down. :(

**Comment by NiR-** [comment]
@dkarlovi We'll release a new version of this repo. From that release we'll recommend developers to require newly created metapackages rather than this repo. So actually we can lock Azure SDK version out there :)

**Comment by dkarlovi** [comment]
@NiR- So you've decided against the mono-repo?

**Comment by NiR-** [comment]
@dkarlovi Currently we're in favor of mono-repo because we think it will be simpler for us to overhaul the project. We might split the repo later though. So yep you're right, it's not ideal because information about supported version is spreaded across many repositories. In the mean we consider it as a little drawback. We can overcome this issue by adding a note in the README + locking the version in `require-dev` section.

Which version introduce the BC break you're talking about? Btw, thank you for reporting it :)

**Comment by dkarlovi** [comment]
@NiR- done, I've squashed and rebased to latest master.

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 4]
```diff
@@ -47,22 +49,52 @@ class AzureBlobStorage implements Adapter,
...
+     */
+    public function getCreateContainerOptions()
+    {
+        return $this->createContainerOptions;
+    }
+
+    /**
+     * @param CreateContainerOptions $options
+     */
+    public function setCreateContainerOptions(CreateContainerOptions $options)
```
Comment (NiR-) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: IMHO it would be better to have this as constructor parameter, it might even replace current `$create` argument: it could either take a `CreateContainerOptions` instance, meaning container(s) should be created if needed, or null, meaning it should fail if container(s) does not exist. But it would be a BC change (actually it would be better in another PR). What do you think?

Comment (dkarlovi) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: I agree actually, did this to keep BC. 👍 

If null, it will be treated as an empty `CreateContainerOptions`, meaning containers get created with whatever the default values other than name are. We can go with this for now and revamp the whole adapter interface with 1.0?

File: src/Gaufrette/Adapter/AzureBlobStorage.php [hunk 5]
```diff
@@ -120,78 +156,91 @@ public function deleteContainer($containerName, DeleteContainerOptions $options
...
 
-            $this->blobProxy->createBlockBlob($this->containerName, $key, $content, $options);
+            $options->setContentType($contentType);
+        }
 
-            if (is_resource($content)) {
-                return Util\Size::fromResource($content);
-            }
+        try {
+            $this->createContainer($containerName);
```
Comment (NiR-) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: Doing a final review I realized this should be done only when `multiContainerMode` is enabled. Also I think `$this->create` should be checked too, so there's no big differences introduced between mono/multi container mode. Beside this point, great job!

File: src/Gaufrette/Adapter/AzureBlobStorage.php (hunk 4 above)
Comment (NiR-) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: Yep, that's what I thought :)

File: src/Gaufrette/Adapter/AzureBlobStorage.php (hunk 5 above)
Comment (dkarlovi) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: Multi-container mode ignores `$create` as it wouldn't make much sense to say "Create containers automatically, but don't". :) Agree to check for multi-mode here, will do it today. 👍 

Thanks, it's for my selfish interest to have this feature, obviously, but glad there's a Gaufrette to contribute it to in the first place, thank you all for that.

Comment (dkarlovi) [inline src/Gaufrette/Adapter/AzureBlobStorage.php]: @NiR- done, also for rename()
